### PR TITLE
fast modular exponentiation implementation

### PR DESCRIPTION
**Pull request by jasonshaev**
Implement https://en.wikipedia.org/wiki/Modular_exponentiation#Right-to-left_binary_method algorithm for fast modular exponentiation on BigNumbers when n > 0 and x, n are integers.

### CODE REVIEW

File: bignumber.js [hunk 1]
```diff
@@ -2369,14 +2369,46 @@
...
+                //Any int modulus 1 is 0
+                if ( m === 1 ) return 0;
+
+                x = x.mod( m );
+                while ( n > 0 ) {
+                    if ( n % 2 === 1 ) {
+                        y = y.times( x ).mod( m );
+                        if ( k && y.c.length > k ) y.c.length = k;
+                    }
+                    n = n >> 1;
```
Comment (MikeMcl) [inline bignumber.js]: Remember `n` can be up a value up to `9007199254740992`, and this bitwise halving will fail for numbers over 2^31, which is why I used `mathfloor( i / 2 )`.

File: bignumber.js [hunk 2]
```diff
@@ -2369,14 +2369,46 @@
...
                 i = mathfloor( n < 0 ? -n : +n ),
-                x = this;
+                x = this,
+                nIsValidInt = isValidInt( n, -MAX_SAFE_INTEGER, MAX_SAFE_INTEGER, 23, 'exponent' ),
+                hasMod = typeof m !== 'undefined';
+
+            //Implement fast modular exponentiation via
+            //RTL binary method when n >= 0 && m > 0
+            //Returns a BigNumber which equals (x ^ n) % m
+            function powMod(x, n, m, k) {
```
Comment (MikeMcl) [inline bignumber.js]: It seem unnecessary to define `powMod` as an inner function when it is only called once. It means that the function may be recreated on each call, and adds extra code needlessly.

**Comment by MikeMcl** [comment]
I've had a quick look through, but have not tested this properly yet. I think the implementation may be a bit bloated. I'd prefer it if you made as few additions as possible and just kept the single calculation loop, adding `mod` conditionally, for example

```
if ( i % 2 ) {
    y = y.times(x);
    if (hasMod) y = y.mod(m); 
    if ( !y.c ) break;
    if ( k && y.c.length > k ) y.c.length = k;
}
```

It means a few more `if` statements in the loop, but their cost is insignificant when compared to operations like `times`.

**Comment by jasonshaev** [comment]
Thanks for you comments. Good point on the bit shift not working for `n > Math.pow(2, 31)`, that had not occurred to me.

Re: the implementation being bloated, I debated whether to try and jam my code into the existing implementation with conditionals versus trying to isolate it in it's own function. I opted for the latter because I thought it would be clearer what code I had touched. However, I realize now that the result is more lines of code and having the function declaration `powMod` where I put it makes no sense, because as you pointed out it is recreated on each call to `pow`, regardless of whether you are using the modular functionality. I put it there simply because I couldn't find a better place for it :)

Will make these changes, thanks for your time!

**Comment by jasonshaev** [comment]
Thanks for your comments, I've updated my pull request. This implementation is MUCH cleaner, thanks for pointing me in the right direction.

**Comment by MikeMcl** [comment]
The method needs to enforce the fact that `x` and `m` must be integers.

Shouldn't there also be

    if (hasMod) x = x.mod(m);

to stop `x` growing very large.

Has the method been tested thoroughly, for example, by testing the return value for equality with the result of `x.pow(n).mod(m)` over many thousands of varying inputs with a variety of `POW_PRECISION` and `DECIMAL_PLACES` settings?

Reliability is the most important attribute of this library.

**Comment by MikeMcl** [comment]
From *v2.3.0*, `toPower` accepts an optional modulus parameter. I didn't use this PR in the end, but the modular exponentiation functionality wouldn't have been added without it.

**Comment by jasonshaev** [comment]
Awesome, glad ot hear the functionality was added!

**Pull request closed without merging**